schema/loopModeEvents.go

```diff
@@ -0,0 +1,98 @@
+package schema
+
+import (
+	"time"
+)
+
+type (
+	LoopModeEvent struct {
+		TimeRange 			`bson:"inline"`
+		DeliveryType string
+	}
+)
+
+func NewLoopModeEvent(start time.Time, end *time.Time, deliveryType string) LoopModeEvent {
+	return LoopModeEvent{
+		TimeRange: TimeRange{
+			Start: start,
+			End: end,
+		},
+		DeliveryType: deliveryType,
+	}
+}
+
+func filterLoopModes(loopModes []LoopModeEvent) []LoopModeEvent {
+	// The terminal sends not updated loopModes without duration
+	// We may have loopMode objects without duration but only one 
+	// i.e. the last one meaning that the terminal is currently in loopMode ON
+	filteredLoopModes := []LoopModeEvent{}
+	lastIndex := len(loopModes) - 1
+	for i, loopMode := range loopModes {
+		if loopMode.End == nil && i != lastIndex {
+			continue
+		}
+		filteredLoopModes = append(filteredLoopModes, loopMode)
+	}
+	return filteredLoopModes
+}
+
+func FillLoopModeEvents(loopModes []LoopModeEvent) []LoopModeEvent {
+	if len(loopModes) == 0 {
+		return []LoopModeEvent{}
+	}
+	filledLoopModes := []LoopModeEvent{}
+	filteredLoopModes := filterLoopModes(loopModes)
+	lastIndex := len(filteredLoopModes) - 1
+	currentLoopMode := filteredLoopModes[0]
+
+	for i := range filteredLoopModes {
+		if i != lastIndex {
+			next := filteredLoopModes[i + 1]
+			unionRanges := currentLoopMode.TimeRange.Union(next.TimeRange, true)
+			lastUnionIndex := len(unionRanges) - 1
+			for j, tRange :=  range unionRanges {
+				deliveryType := "automated"
+				if tRange.Fill {
+					deliveryType = "scheduled"
+				}
+				loopModeEvent := NewLoopModeEvent(tRange.Start, tRange.End, deliveryType)
+				if j == lastUnionIndex {
+					currentLoopMode = loopModeEvent
+				} else {
+					filledLoopModes = append(filledLoopModes, loopModeEvent)
+				}
+			}
+		} else {
+			filledLoopModes = append(filledLoopModes, currentLoopMode)
+		}
+	}
+	return filledLoopModes
+}
+
+func GetLoopModeEventsBetween(start time.Time, end * time.Time, loopModes []LoopModeEvent) []LoopModeEvent {
+	matchedLoopModes := []LoopModeEvent{}
+	myLoopMode  := NewLoopModeEvent(start, end, "")
+	for _, loopMode := range loopModes {
+		if loopMode.TimeRange.OverLaps(myLoopMode.TimeRange) {	
+			matchingLoopMode :=  LoopModeEvent{
+				TimeRange{
+					Start: loopMode.Start,
+					End: loopMode.End,
+				},
+				loopMode.DeliveryType,
+			}
+			if matchingLoopMode.Start.Before(myLoopMode.Start) {
+				matchingLoopMode.Start = myLoopMode.Start
+			}
+			if TimeAfterOrEqual(matchingLoopMode.End, myLoopMode.End) {
+				matchingLoopMode.End = myLoopMode.End
+			}
+			matchedLoopModes = append(matchedLoopModes, matchingLoopMode)
+		}
+	}
+	if len(matchedLoopModes) == 0 {
+		myLoopMode.DeliveryType = "scheduled"
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❔ question:** ‏ Why is the basal scheduled when there is no matching ? I would expect manual ...